README


This archive contains the distribution Wx-App-AnnualCal,
version 0.91:

  the main module of the AnnualCal distribution

This software is copyright (c) 2014 by Elliot Winston.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.015.



Changes
0.91      2014-04-28 22:56:06EDT-0400 America/New_York
  - corrected package names

0.9       2014-04-21 12:01:59EDT-0400 America/New_York
  - initial release


lib/Wx/App/AnnualCal.pm
package Wx::App::AnnualCal;

use strict;
use warnings;

use Wx 0.990 qw(:frame);
use base qw(Wx::App);

use lib qw(../../../lib);
use Wx::App::AnnualCal::MyFrame;

my $VERSION = 0.91;

          ##################################################

sub OnInit
  {
  my $frame = Wx::App::AnnualCal::MyFrame->new
    (
     undef,                                 # parent window
     -1,                                    # default id value
     'Annual Calendar',                     # title
     [-1,-1],                               # default position
     [-1,-1],                               # default size
     wxDEFAULT_FRAME_STYLE,                 # frame style
    );                                   
  $frame->build();                          # creates calendar
  $frame->Center(wxBOTH);                   # center frame on screen
  $frame->Show(1);                          # displays calendar

  return(1);
  }

# * * *
#
#  Input for Dist::Zilla::Pod::Weaver plugin to create POD documentation.
#
# * * *

#ABSTRACT: the main module of the AnnualCal distribution


1;


__END__
=pod

=head1 NAME

Wx::App::AnnualCal - the main module of the AnnualCal distribution

=head1 VERSION

version 0.91

=head1 SYNOPSIS

Main module for AnnualCal, a GUI application programmed in wxPerl,
displaying an annual calendar for the year input by the user.

=head1 METHODS

=head2 OnInit

The entry point to a wxPerl program, analagous to 'main' in a C program.

=head1 DESIGN

The design of the GUI is moderately complex, utilizing BoxSizer, Grid,
GridSizer, TxtCtrl, and Button widgets.

The name of each month is inserted into a TxtCtrl, and the days of each
month are inserted into a 7x7 GridSizer, the first row of which is reserved
for the weekday names.
Each of these 12 pairs of widgets is then inserted into a vertical BoxSizer,
the TxtCtrl added prior to the GridSizer, creating a set of 12 widgets
containing all the data needed to display a year.  The next step is to insert 
this collection into a single 3x4 Grid, in the correct order; a GridSizer
is not used here because today's date is highlighted, when it is displayed.

A horizontal BoxSizer is used to contain the interactive
Buttons for changing the year, and a TxtCtrl to display the current year.

Finally, a vertical BoxSizer is created which contains the Grid on top, and
the horizontal BoxSizer on the bottom.  The motivation for this design is
that only the 'upper sizer' has to be replaced when the user changes the
year; the 'bottom sizer' is essentially static, except for updating
the TexCtrl to display the new year.

=head1 AUTHOR

Elliot Winston <exw@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Elliot Winston.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Wx/App/AnnualCal/MyFrame.pm
package Wx::App::AnnualCal::MyFrame;

use strict;
use warnings;
use feature qw(switch);

use Wx 0.990 qw(wxICON_ERROR wxTheApp wxID_OK
                :font :sizer :dialog :textctrl :button);
use Wx::Event qw(EVT_BUTTON);
use base qw(Wx::Frame);

use Date::Calc 6.3 qw(Today Month_to_Text);
use Readonly 1.03;

use lib qw(../../../../lib);
use Wx::App::AnnualCal::MonthSizer;

          ##################################################

sub new
  {
  my $class = shift;
  my $self = $class->SUPER::new(@_);

  my $panel = Wx::Panel->new($self);
  $panel->SetBackgroundColour(Wx::Colour->new('black'));

  my ($yr, $mon, $day) = Today(0);
  my $current = {'day' => $day, 'month' => Month_to_Text($mon), 'year' => $yr};

  my $day_names = ['Sun','Mon','Tue','Wed','Thu','Fri','Sat'];

  my $fonts = {'norm' => Wx::Font->new(9, wxFONTFAMILY_MODERN,
                           wxFONTSTYLE_NORMAL, wxFONTWEIGHT_NORMAL),
               'ital' => Wx::Font->new(9, wxFONTFAMILY_MODERN,
                           wxFONTSTYLE_ITALIC, wxFONTWEIGHT_BOLD),
               'emph' => Wx::Font->new(10, wxFONTFAMILY_MODERN,
                           wxFONTSTYLE_ITALIC, wxFONTWEIGHT_BOLD),
               'bold' => Wx::Font->new(12, wxFONTFAMILY_MODERN,
                           wxFONTSTYLE_NORMAL, wxFONTWEIGHT_BOLD),
              };

  my $colors = {'gold' => Wx::Colour->new(255,215,0),
                'gray' => Wx::Colour->new(100,100,100)};

  my $year = $ARGV[0] || $current->{year};
  if ($year =~ /\A[0-9]+\Z/x and 1 <= $year and $year <= 32767)
    {
    chomp($year);
    }
  else
    {
    Wx::MessageBox('Enter a year between 1 and 32767.', 'ERROR',
                   wxICON_ERROR|wxOK);
    exit(0);
    }

  $self->{param} = {'current' => $current, 'day_names' => $day_names,
                    'fonts' => $fonts, 'panel' => $panel, 'year' => $year,
                    'colors' => $colors};

  Readonly my %IDS => (PRIOR_BUTTON => 1001,
                       NEXT_BUTTON  => 1002,
                       ANY_BUTTON   => 1003,
                       EXIT_BUTTON  => 1004,
                      );

  my $priorbtn = Wx::Button->new($panel,            # parent window
                                 $IDS{PRIOR_BUTTON},   # id
                                 ' < ',             # label
                                 [-1,-1],           # default location
                                 [-1,-1],           # default size
                                 wxBU_EXACTFIT,     # minimum size
                                );
  $self->{priorbtn} = $priorbtn;
  $priorbtn->SetFont($fonts->{bold});
  $priorbtn->SetForegroundColour(Wx::Colour->new('green'));
  $priorbtn->SetBackgroundColour($colors->{'gray'});
  ($year == 1) ? $priorbtn->Enable(0) : $priorbtn->Enable(1);
  EVT_BUTTON($self, $IDS{PRIOR_BUTTON}, \&ClickPRIOR);

  my $yeartxt = Wx::TextCtrl->new($panel,
                                  -1,
                                  $year,
                                  [-1,-1],
                                  [-1,-1],
                                  wxTE_CENTRE|wxTE_READONLY,
                                 );
  $yeartxt->SetFont($fonts->{bold});
  $yeartxt->SetForegroundColour(Wx::Colour->new('black'));
  $yeartxt->SetBackgroundColour($colors->{'gold'});
  $self->{yeartxt} = $yeartxt;

  my $nextbtn = Wx::Button->new($panel,             # parent window
                                $IDS{NEXT_BUTTON},     # id
                                ' > ',              # label
                                 [-1,-1],           # default location
                                 [-1,-1],           # default size
                                 wxBU_EXACTFIT,     # minimum size
                               );
  $self->{nextbtn} = $nextbtn;
  $nextbtn->SetFont($fonts->{bold});
  $nextbtn->SetForegroundColour(Wx::Colour->new('green'));
  $nextbtn->SetBackgroundColour($colors->{'gray'});
  ($year == 32767) ? $nextbtn->Enable(0) : $nextbtn->Enable(1);
  EVT_BUTTON($self, $IDS{NEXT_BUTTON}, \&ClickNEXT);

  my $anybtn = Wx::Button->new($panel,              # parent window
                               $IDS{ANY_BUTTON},       # id
                               'ANY YEAR',          # label
                              );
  $anybtn->SetFont($fonts->{bold});
  $anybtn->SetForegroundColour(Wx::Colour->new('green'));
  $anybtn->SetBackgroundColour($colors->{'gray'});
  EVT_BUTTON($self, $IDS{ANY_BUTTON}, \&ClickANY);

  my $exitbtn = Wx::Button->new($panel,           # parent window
                                $IDS{EXIT_BUTTON},   # id
                                'EXIT'            # label
                               );
  $exitbtn->SetFont($fonts->{bold});
  $exitbtn->SetForegroundColour(Wx::Colour->new('white'));
  $exitbtn->SetBackgroundColour(Wx::Colour->new('red'));
  EVT_BUTTON($self, $IDS{EXIT_BUTTON}, sub {wxTheApp->ExitMainLoop()});

  my $lowersizer = Wx::BoxSizer->new(wxHORIZONTAL);
  $lowersizer->AddSpacer(20);
  $lowersizer->Add($anybtn,             # button control
                   1,                   # unit length
                   wxBOTTOM|            # bottom border
                   wxALIGN_CENTER,      # central alignment
                   20,                  # border width
                  );
  $lowersizer->AddSpacer(30);
  $lowersizer->Add($priorbtn,           # button control
                   0,                   # exact fit
                   wxBOTTOM|            # bottom border
                   wxALIGN_CENTER,      # central alignment
                   20,                  # border width
                  );
  $lowersizer->Add($yeartxt,            # text control
                   1.5,                 # proportional length
                   wxBOTTOM|            # bottom border
                   wxALIGN_CENTER,      # central alignment
                   20,                  # border width
                  );
  $lowersizer->Add($nextbtn,            # button control
                   0,                   # exact fit
                   wxBOTTOM|            # bottom border
                   wxALIGN_CENTER,      # central alignment
                   20,                  # border width
                  );
  $lowersizer->AddSpacer(30);
  $lowersizer->Add($exitbtn,            # button control
                   1,                   # unit length
                   wxBOTTOM|            # bottom border
                   wxALIGN_CENTER,      # central alignment
                   20,                  # border width
                  );
  $lowersizer->AddSpacer(20);

  $self->{lowersizer} = $lowersizer;

  return($self);
  }

          ##################################################

sub build
  {
  my $self = shift;
  my $panel = $self->{param}->{panel};
  my $year = $self->{param}->{year};

  my $ms = Wx::App::AnnualCal::MonthSizer->new($self->{param});

  my $gridsizer = Wx::GridSizer->new(3,4,0,0);
  map { $gridsizer->Add($ms->getmonthsizer($_)) } (0..11);
  $self->{gridsizer} = $gridsizer;

  my $margin = Wx::Panel->new($self, -1, [-1,-1], [13,-1]);
  $margin->SetBackgroundColour(Wx::Colour->new('black'));
  $self->{margin} = $margin;

  my $uppersizer = Wx::BoxSizer->new(wxHORIZONTAL);
  $uppersizer->Add($margin);
  $uppersizer->Add($gridsizer);
  $self->{uppersizer} = $uppersizer;

  my $panelsizer = Wx::BoxSizer->new(wxVERTICAL);
  $panelsizer->Add($uppersizer);
  $panelsizer->Add($self->{lowersizer},1,wxGROW);

  $panel->SetSizer($panelsizer);

  my $framesizer = Wx::BoxSizer->new(wxVERTICAL);
  $framesizer->Add($panel);
  $self->SetSizerAndFit($framesizer);

  return(1);
  }

          ##################################################

sub update
  {
  my $self = shift;

  my $ms = Wx::App::AnnualCal::MonthSizer->new($self->{param});

  my $gridsizer = Wx::GridSizer->new(3,4,0,0);
  map { $gridsizer->Add($ms->getmonthsizer($_)) } (0..11);

  my $uppersizer = $self->{uppersizer};
  $uppersizer->Detach(1);
  $uppersizer->Add($gridsizer);
  $uppersizer->Layout();

  my $yeartxt = $self->{yeartxt};
  $yeartxt->Clear();
  $yeartxt->ChangeValue($self->{param}->{year});

  return(1);
  }

          ##################################################

sub ClickPRIOR
  {
  my $self = shift;
  my $year = $self->{param}->{year} - 1;
  $self->{param}->{year} = $year;

  $self->{nextbtn}->Enable(1);
  ($year == 1) ? $self->{priorbtn}->Enable(0) : $self->{priorbtn}->Enable(1);

  $self->update();
  return;
  }

          ##################################################

sub ClickNEXT
  {
  my $self = shift;
  my $year = $self->{param}->{year} + 1;
  $self->{param}->{year} = $year;

  $self->{priorbtn}->Enable(1);
  ($year == 32767) ? $self->{nextbtn}->Enable(0) : $self->{nextbtn}->Enable(1);

  $self->update();
  return;
  }

          ##################################################

sub ClickANY
  {
  my $self = shift;
  my $dlg = Wx::TextEntryDialog->new($self, 'Enter a year between 1 and 32767.',
                                    'USER INPUT', '', wxOK|wxCANCEL);
  my $ans = $dlg->ShowModal();
  if ($ans == wxID_OK)
    {
    my $year = $dlg->GetValue();
    if ($year =~ /\A[0-9]+\Z/x and 1 <= $year and $year <= 32767)
      {
      $self->{param}->{year} = $year;
      }
    else
      {
      $dlg->Destroy();
      Wx::MessageBox("$year is invalid.", 'ERROR', wxICON_ERROR|wxOK);
      }
    }
  else
    {
    $dlg->Destroy();
    }

  my $year = $self->{param}->{year};
  given ($year)
    {
    when (1)
      {
      $self->{priorbtn}->Enable(0);
      $self->{nextbtn}->Enable(1);
      }
    when (32767)
      {
      $self->{priorbtn}->Enable(1);
      $self->{nextbtn}->Enable(0);
      }
    default
      {
      $self->{nextbtn}->Enable(1);
      $self->{priorbtn}->Enable(1);
      }
    }
  $self->update();
  return;
  }

# * * *
#
#  Input for Dist::Zilla::Pod::Weaver plugin to create POD documentation.
#
# * * *

#ABSTRACT: a module in the AnnualCal distribution

1;

__END__
=pod

=head1 NAME

Wx::App::AnnualCal::MyFrame - a module in the AnnualCal distribution

=head1 VERSION

version 0.91

=head1 SYNOPSIS

Package which defines, lays out, and maintains the widgets in the GUI.

=head1 METHODS

=head2 new

Constructor defining all the GUI parameters and basic widgets,
and builds the 'lower sizer' (see L<Wx::App::AnnualCal/DESIGN>).

=head2 build

Method which gets and/or builds the higher-order widgets which are ultimately
inserted into the panel and frame for display.

=head2 update

Method which responds to a user request for a new year
by getting a new 'upper sizer' (see L<Wx::App::AnnualCal/DESIGN>),
and replacing the old one.
Also, the display of the numeric new year is updated in the 'lower sizer'.

=head2 ClickPRIOR

PRIOR Button callback which resets the 'year' parameter to the prior year
and calls the 'update' method.

=head2 ClickNEXT

NEXT Button callback which resets the 'year' parameter to the next year
and calls the 'update' method.

=head2 ClickANY

ANY Button callback which fetches the input year from the TextEntryDialog
popup and calls the 'update' method.

=head1 AUTHOR

Elliot Winston <exw@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Elliot Winston.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Wx/App/AnnualCal/MonthSizer.pm
package Wx::App::AnnualCal::MonthSizer;

use strict;
use warnings;

use Wx 0.990 qw(:sizer :textctrl);
use Wx::Grid;

use Date::Calc 6.3 qw(Day_of_Week Days_in_Month);

          ##################################################

sub new
  {
  my ($class, $param) = @_;

  my $year = $param->{year};

  my $months = [{'name' => 'JANUARY',   'len' => 31, 'days' => []},
                {'name' => 'FEBRUARY',  'len' => Days_in_Month($year,2), 'days' => []},
                {'name' => 'MARCH',     'len' => 31, 'days' => []},
                {'name' => 'APRIL',     'len' => 30, 'days' => []},
                {'name' => 'MAY',       'len' => 31, 'days' => []},
                {'name' => 'JUNE',      'len' => 30, 'days' => []},
                {'name' => 'JULY',      'len' => 31, 'days' => []},
                {'name' => 'AUGUST',    'len' => 31, 'days' => []},
                {'name' => 'SEPTEMBER', 'len' => 30, 'days' => []},
                {'name' => 'OCTOBER',   'len' => 31, 'days' => []},
                {'name' => 'NOVEMBER',  'len' => 30, 'days' => []},
                {'name' => 'DECEMBER',  'len' => 31, 'days' => []},
               ];

  for (my $i=0; $i<12; $i++)
    {
    my $col = Day_of_Week($year, $i+1, 1);
    $col %= 7;
    my $row = 1;
    for (my $j=1; $j<=$months->[$i]{len}; $j++)
      {
      $months->[$i]{days}->[$j] = {'row' => $row, 'col' => $col};
      if ($col == 6)
        {
        $col = 0;
        $row++;
        }
      else
        {
        $col++;
        }
      }
    }

  my $self = {'months' => $months, 'param' => $param};

  bless($self, $class);
  return($self);
  }

          ##################################################

sub getmonthsizer
  {
  my ($self, $ind) = @_;

  my $months = $self->{months};
  my $ref = $self->{param};
  my ($panel, $year, $fonts, $current, $day_names, $colors) =
     @{$ref}{qw(panel year fonts current day_names colors)};

  my $txt = Wx::TextCtrl->new($panel,
                              -1,
                              $months->[$ind]{name},
                              [-1,-1],
                              [-1,-1],
                              wxTE_CENTRE|wxTE_READONLY,
                             );
  $txt->SetFont($fonts->{ital});
  $txt->SetForegroundColour(Wx::Colour->new('green'));
  $txt->SetBackgroundColour(Wx::Colour->new('black'));

  my $grid = Wx::Grid->new($panel, -1, [-1,-1]);
  my ($nrow, $ncol) = (7,7);
  $grid->CreateGrid($nrow, $ncol);
  $grid->SetRowLabelSize(0);
  $grid->SetColLabelSize(0);
  $grid->SetGridLineColour(Wx::Colour->new('black'));
  $grid->SetDefaultCellBackgroundColour(Wx::Colour->new('black'));
  $grid->SetDefaultCellTextColour($colors->{'gold'});
  $grid->SetDefaultCellFont($fonts->{norm});
  $grid->SetDefaultCellAlignment(wxALIGN_CENTRE, wxALIGN_CENTRE);
  $grid->AutoSize();
  my $cell_wide = $grid->GetColSize(0) + 2;

  for (my $j=0; $j<$ncol; $j++)
    {
    $grid->SetColSize($j, $cell_wide);
    $grid->SetCellValue(0, $j, $day_names->[$j]);
    $grid->SetCellFont(0, $j, $fonts->{norm});
    };
  for (my $j=0; $j<$nrow; $j++)
    {
    $grid->SetCellTextColour($j, 0, Wx::Colour->new('cyan'));
    $grid->SetCellTextColour($j, $ncol-1, Wx::Colour->new('cyan'));
    }

  my $ndays = $months->[$ind]{len};
  $ref = $months->[$ind]{days};
  for (my $day=1; $day<=$ndays; $day++)
    {
    my $row = $ref->[$day]{row};
    my $col = $ref->[$day]{col};
    $grid->SetCellValue($row, $col, $day);
    }

  if ($year == $current->{year})
    {
    if (uc($current->{month}) eq $months->[$ind]{name})
      {
      my $row = $ref->[$current->{day}]{row};
      my $col = $ref->[$current->{day}]{col};
      $grid->SetCellTextColour($row, $col, Wx::Colour->new('black'));
      $grid->SetCellBackgroundColour($row, $col, $colors->{'gold'});
      $grid->SetCellFont($row, $col, $fonts->{emph});
      }
    }

  my $monthsizer = Wx::BoxSizer->new(wxVERTICAL);
  $monthsizer->Add($txt,              # text control
                   0,                 # vertically unstretchable
                   wxALIGN_CENTER,    # central alignment
                  );                    
  $monthsizer->Add($grid);            

  return($monthsizer);
  }

# * * *
#
#  Input for Dist::Zilla::Pod::Weaver plugin to create POD documentation.
#
# * * *

#ABSTRACT: a module in the AnnualCal distribution

1;

__END__
=pod

=head1 NAME

Wx::App::AnnualCal::MonthSizer - a module in the AnnualCal distribution

=head1 VERSION

version 0.91

=head1 SYNOPSIS

Package which generates and stores the required data for the display.

=head1 METHODS

=head2 new

Given the year to be displayed, which is passed as a value in a data
container, the constructor populates the container with additional
data describing the days of each month of the year.

=head2 getmonthsizer

Method which creates and returns the 'upper sizer' which displays the calendar
for the desired year (see L<Wx::App::AnnualCal/DESIGN>).

=head1 AUTHOR

Elliot Winston <exw@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Elliot Winston.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "the main module of the AnnualCal distribution",
  "AUTHOR" => "Elliot Winston <exw\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Wx-App-AnnualCal",
  "EXE_FILES" => [
    "bin/AnnualCal.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "Wx::App::AnnualCal",
  "PREREQ_PM" => {
    "Date::Calc" => "6.3",
    "Readonly" => "1.03",
    "Wx" => "0.990",
    "Wx::App" => 0,
    "Wx::Grid" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "0.91",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Date::Calc" => "6.3",
  "Readonly" => "1.03",
  "Test::More" => 0,
  "Wx" => "0.990",
  "Wx::App" => 0,
  "Wx::Grid" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



